Firestore/core/test/unit/credentials/firebase_app_check_credentials_provider_test.mm

```diff
@@ -178,6 +178,44 @@ - (nonnull NSString*)tokenDidChangeNotificationName {
   ASSERT_EQ(std::future_status::ready, token_future.wait_for(kTimeout));
 }
 
+// Regression test for https://github.com/firebase/firebase-ios-sdk/issues/8895
+TEST(FirebaseAppCheckCredentialsProviderTest,
+     ListenForTokenChangesIgnoresUnrelatedNotifcations) {
+  auto token_promise = std::make_shared<std::promise<std::string>>();
+
+  FIRApp* app = testutil::AppForUnitTesting();
+  FSTAppCheckFake* app_check = [[FSTAppCheckFake alloc] initWithToken:@""];
+  FirebaseAppCheckCredentialsProvider credentials_provider(app, app_check);
+
+  credentials_provider.SetCredentialChangeListener(
+      [token_promise](const std::string& result) {
+        if (result != "") {
+          EXPECT_EQ("token2", result);
+          token_promise->set_value(result);
+        }
+      });
+
+  [[NSNotificationCenter defaultCenter]
+      postNotificationName:@"unrelated"
+                    object:app_check
+                  userInfo:@{
+                    [app_check notificationTokenKey] : @"token1",
+                    [app_check notificationAppNameKey] : [app name],
+                  }];
+
+  [[NSNotificationCenter defaultCenter]
+      postNotificationName:[app_check tokenDidChangeNotificationName]
+                    object:app_check
+                  userInfo:@{
+                    [app_check notificationTokenKey] : @"token2",
+                    [app_check notificationAppNameKey] : [app name],
+                  }];
+
+  auto kTimeout = std::chrono::seconds(5);
+  auto token_future = token_promise->get_future();
+  ASSERT_EQ(std::future_status::ready, token_future.wait_for(kTimeout));
```

Comment:
The test would fail in line 193, as the comparisons of the token value would fail. The promise does only resolve once, but it is merely used to ensure that the callback fired. 
I moved the assert of the token value out of the callback - maybe that makes it a bit cleaner.

Reply:
Hm... It looks like if `@"unrelated"` notification passes through then the promise value will be still overwritten to `"token2"` as it is called later, so the test won't actually fail. Am I missing anything?

Reply:
Promises can only be resolved once.

Reply:
Ah, sorry, I misread the comment above!